### PR TITLE
Update pin memory

### LINKED ISSUE #7497: D-FINE-M training with tiling sometimes fails during OV evaluation
**Describe the bug**

Training a model on a large dataset with tiling enabled failed during OV model evaluation, despite abundant HW resources on my server (128GB RAM and B70 32GB GPU).

**Steps to Reproduce**

1. Import a large detection dataset, e.g. [this dataset](https://intel-my.sharepoint.com/:u:/p/leonardo_lai/IQDns4bFftR6SodBwC4cUbg_AUv88Wnnw4Rf9ZIJFp6RaRo?e=iUqn3P)
2. Train D-FINE-M with tiling enabled (automatic). Optionally, reduce the epochs to speed things up.
3. Monitor the training job -> it fails at the end

Image

Full log: [dfine-training.log](https://github.com/user-attachments/files/31839707/dfine-training.log)

**Environment:**

- Geti app / library version: v3.2rc0
- Installation mode: Docker
- Hardware spec:
  - Intel Core i9-13900K 
  - 128GB RAM
  - Intel ARC B70 GPU (32GB)
- OS: Ubuntu 24.04

### LINKED ISSUE #7507: MaskRCNN SwinT training sometimes fails during OV evaluation
**Describe the bug**

Training a MaskRCNN SwinT model on an instance segmentation project failed during the evaluation of the OpenVINO model, with no visible error in the log.

**Steps to Reproduce**

1. Create an instance segmentation project, e.g. importing [this dataset](https://intel-my.sharepoint.com/:u:/p/leonardo_lai/IQAli9e4zB4FQaf-jsTisqXJAeUupMFGf6uelC28kvOV3Zs?e=Ghh7J3)
2. Train a model of type Mask-RCNN SwinT, default settings
3. Monitor the training job -> it fails

**Full log:** [maskrcnn-swint.log](https://github.com/user-attachments/files/31899206/maskrcnn-swint.log)

No errors, the last lines in the log are:
```
2026-09-06 13:57:08.605 | INFO     | app.execution.training.getitune_trainer:evaluate_model:562 - Evaluating the openvino model...
2026-09-06 13:57:08.844 | INFO     | getitune.cli.utils.jsonargparse:get_configuration:451 - /application/backend/.venv/lib/python3.13/site-packages/getitune/recipe/instance_segmentation/rfdetr_seg_small.yaml is loaded.
2026-09-06 13:57:08.853 | INFO     | model_api.adapters.openvino_adapter:create_core:56 - OpenVINO Runtime
2026-09-06 13:57:08.854 | INFO     | model_api.adapters.openvino_adapter:create_core:57 -       build: 2026.3.1-22476-759c5a6ab8c-releases/2026/3
2026-09-06 13:57:08.854 | INFO     | model_api.adapters.openvino_adapter:__init__:180 - Reading model /application/data/getitune-workspace-3a306236-723a-4818-82c6-bd469eb096d5/exported_model.xml
2026-09-06 13:57:10.200 | INFO     | model_api.adapters.openvino_adapter:load_model:247 - The model /application/data/getitune-workspace-3a306236-723a-4818-82c6-bd469eb096d5/exported_model.xml is loaded to CPU
2026-09-06 13:57:10.201 | INFO     | model_api.adapters.openvino_adapter:log_runtime_settings:274 -     Number of model infer requests: 16
2026-09-06 13:57:11.777 | INFO     | model_api.adapters.openvino_adapter:load_model:247 - The model /application/data/getitune-workspace-3a306236-723a-4818-82c6-bd469eb096d5/exported_model.xml is loaded to CPU
2026-09-06 13:57:11.779 | INFO     | model_api.adapters.openvino_adapter:log_runtime_settings:274 -     Number of model infer requests: 16
2026-09-06 13:57:11.865 | INFO     | getitune.cli.utils.jsonargparse:get_configuration:451 - /application/backend/.venv/lib/python3.13/site-packages/getitune/recipe/instance_segmentation/openvino_model.yaml is loaded.
2026-09-06 13:57:11.865 | WARNING  | getitune.data.module:from_vision_datasets:327 - The provided train SubsetConfig contains augmentations_cpu which will be overridden by the transforms of the provided VisionDataset. When building DataModule from pre-constructed datasets, developers should set up the transforms when creating the datasets.
2026-09-06 13:57:11.865 | WARNING  | getitune.data.module:from_vision_datasets:327 - The provided val SubsetConfig contains augmentations_cpu which will be overridden by the transforms of the provided VisionDataset. When building DataModule from pre-constructed datasets, developers should set up the transforms when creating the datasets.
2026-09-06 13:57:11.865 | WARNING  | getitune.data.module:from_vision_datasets:327 - The provided test SubsetConfig contains augmentations_cpu which will be overridden by the transforms of the provided VisionDataset. When building DataModule from pre-constructed datasets, developers should set up the transforms when creating the datasets.
2026-09-06 13:57:11.865 | INFO     | getitune.data.augmentation.intensity:build_intensity_transform:312 - Built intensity transform pipeline: [ScaleToUnit(max_value=255.0)]
```

**Environment:**

- Geti app / library version: v3.2rc0
- Installation mode: Docker
- Hardware spec:
  - Intel Core i9-13900K 
  - 128GB RAM
  - Intel ARC B70 GPU (32GB)
- OS: Ubuntu 24.04

### PR DESCRIPTION
**Pull request by A-Artemis**
### Summary

Memory from training stays pinned during evaluation
- The eval dataloader still pins into accelerator memory. _eval_loader_kwargs hard-coded pin_memory=True; PyTorch pins via torch.accelerator.current_accelerator(), on XPU that is Level Zero host-pinned memory, allocated by a dedicated thread, for a workload that never touches the GPU. update_ov_subset_pipeline propagated the training device into the rebuilt datamodule.

### Checklist

- [x] The PR title and description are clear and descriptive
- [x] I have manually tested the changes
- [x] All changes are covered by automated tests
- [x] All related issues are linked to this PR (if applicable)
- [ ] Documentation has been updated (if applicable)

### CODE REVIEW

File: application/backend/uv.lock
Comment (leoll2) [inline application/backend/uv.lock:1]: Why update `uv.lock` if there is no change in `pyproject.toml`?

File: application/docker/Dockerfile [hunk 1]
```diff
@@ -453,8 +453,18 @@ USER root
 # Includes Intel graphics drivers (steps from https://dgpu-docs.intel.com/driver/client/overview.html#ubuntu-latest)
 SHELL ["/bin/bash", "-o", "pipefail", "-c"]
 
-RUN getent group render_host >/dev/null || groupadd -g 992 render_host && usermod -aG render_host "${USER_NAME}"
-
+ARG RENDER_GID=992
+RUN if ! getent group render_host >/dev/null; then \
+        groupadd -g "${RENDER_GID}" render_host || groupadd render_host; \
+    fi && \
+    usermod -aG render_host "${USER_NAME}"
```
Comment (leoll2) [inline application/docker/Dockerfile]: I don't understand the purpose of this fix. If it's a separate bug, let's track it with a different ticket and branch, so that this PR can remain focused on the high-exposure memory issue.

File: library/src/getitune/data/module.py [hunk 2]
```diff
@@ -513,12 +513,28 @@ def _eval_loader_kwargs(self, dataset: VisionDataset, config: SubsetConfig) -> d
...
         }
 
+    @property
+    def _pin_memory(self) -> bool:
+        """Whether batches should be copied into accelerator-pinned host memory.
+
+        ``DataLoader(pin_memory=True)`` does not pin into ordinary host memory: it
+        pins into memory owned by the *current accelerator* (CUDA / XPU), from a
+        dedicated pinning thread. That is only worth doing when the batches are
+        actually going to be copied to that accelerator.
```
Comment (leoll2) [inline library/src/getitune/data/module.py]: This sentence is a bit confusing as it's currently phrased. In fact, pinned memory is still _host_ memory, the difference is that it's reserved for the accelerator and it cannot be swapped like regular memory pages.

File: library/src/getitune/data/module.py [hunk 3]
```diff
@@ -513,12 +513,28 @@ def _eval_loader_kwargs(self, dataset: VisionDataset, config: SubsetConfig) -> d
...
+
+        For CPU-only consumers - most importantly the OpenVINO evaluation pipeline,
+        which always runs on CPU - pinning adds an extra copy per batch and, worse,
+        makes a CPU-only workload depend on the GPU runtime (Level Zero on XPU,
+        CUDA on NVIDIA) for a large, non-pageable allocation. Keep it off in that case.
+        """
+        return self.device not in (DeviceType.cpu,)
```
Comment (leoll2) [inline library/src/getitune/data/module.py]: Why not simply `self.device != DeviceType.cpu`?

File: application/backend/app/core/jobs/exec/process_run.py [hunk 4]
```diff
@@ -78,10 +79,43 @@ def events(self) -> Iterator[ExecutionEvent]:
...
+            signal_name = signal.Signals(-code).name
+        except ValueError:
+            signal_name = f"signal {-code}"
+
+        hints = {
+            "SIGKILL": (
+                "the process was killed, most likely by the out-of-memory killer; "
+                "check the host memory available to the container"
+            ),
+            "SIGSEGV": "the process crashed in native code (segmentation fault)",
```
Comment (leoll2) [inline application/backend/app/core/jobs/exec/process_run.py]: What do you mean by "native code"?

File: application/backend/app/execution/training/getitune_trainer.py [hunk 5]
```diff
@@ -594,6 +601,14 @@ def evaluate_model(
...
                 if variant.format == ModelFormat.PYTORCH:
                     pytorch_metrics = metrics
+                    # The trained model, optimizer state and the trainer's dataloader
+                    # workers are no longer needed, release them for a clean evaluation env.
+                    self._release_accelerator_memory()
+            finally:
+                # Drop the per-variant engine *before* the next variant compiles its own.
+                if engine is not None and engine is not getitune_engine:
+                    del engine
+                    gc.collect()
```
Comment (leoll2) [inline application/backend/app/execution/training/getitune_trainer.py]: I don't understand the need to manually call the garbage collector. Normally, when a Python object stops being referenced, the Python runtime automatically cleans it up and frees the resources. If you see that the object remains in memory longer than expected, I'd rather investigate if any other object keeps it alive through a reference.

File: application/backend/uv.lock
Comment (A-Artemis) [reply application/backend/uv.lock:1]: This is a leftover from when I was using the a git commit for installing MAPI, it do need revert it back,

File: application/backend/app/core/jobs/exec/process_run.py (hunk 4 above)
Comment (A-Artemis) [reply application/backend/app/core/jobs/exec/process_run.py]: Code that is outside of the main Python code, which could be the any from within the libraries such as C++, ...

I will just make the comment simpler.

File: application/docker/Dockerfile [hunk 6]
```diff
@@ -455,6 +455,12 @@ SHELL ["/bin/bash", "-o", "pipefail", "-c"]
 
 RUN getent group render_host >/dev/null || groupadd -g 992 render_host && usermod -aG render_host "${USER_NAME}"
 
+# WARNING: this is an Ubuntu 'noble' PPA being added to a Debian 13 (trixie) base image,
+# with signature verification disabled (trusted=yes) and no version pins. That makes the
+# resulting driver stack both unverified and non-reproducible: two builds of the same
+# commit can end up with different Level Zero / compute-runtime versions, and apt is free
+# to satisfy individual dependencies from Debian instead of the PPA.
+# TODO: switch to the signed Intel graphics repository for Debian and pin every package.
```
Comment (leoll2) [inline application/docker/Dockerfile]: Can you revert this comment too? It's not wrong per se, but the steps are consistent with the [official instructions](https://dgpu-docs.intel.com/installation-guides/installing-packages-from-the-intel-ppa.html) to install XPU drivers. There is no signed repo to switch to, as far as I know.

**Review by kprokofi — APPROVED** [review]

**Review by leoll2 — APPROVED** [review]
